openhcl/openhcl_boot/src/hypercall.rs

```diff
@@ -85,47 +67,59 @@ impl HvCall {
     /// necessary.
     #[cfg(target_arch = "x86_64")]
     pub fn hypercall_page(&mut self) -> u64 {
-        self.init_if_needed();
+        assert!(self.initialized);
         core::ptr::addr_of!(minimal_rt::arch::hypercall::HYPERCALL_PAGE) as u64
     }
 
-    fn init_if_needed(&mut self) {
+    /// Hypercall initialization.
+    pub fn initialize(&mut self, isolation_type: IsolationType) {
         if !self.initialized {
-            self.initialize();
+            match isolation_type {
+                #[cfg(target_arch = "x86_64")]
+                IsolationType::Tdx => {
+                    self.initialize_tdx();
+                }
+                #[cfg(target_arch = "x86_64")]
+                IsolationType::Snp => (),
+                _ => {
+                    // TODO: revisit os id value. For now, use 1 (which is what UEFI does)
+                    let guest_os_id = hvdef::hypercall::HvGuestOsMicrosoft::new().with_os_id(1);
+
+                    crate::arch::hypercall::initialize(guest_os_id.into());
+                }
+            };
         }
-    }
-
-    pub fn initialize(&mut self) {
-        assert!(!self.initialized);
-
-        // TODO: revisit os id value. For now, use 1 (which is what UEFI does)
-        let guest_os_id = hvdef::hypercall::HvGuestOsMicrosoft::new().with_os_id(1);
-        crate::arch::hypercall::initialize(guest_os_id.into());
-
         self.initialized = true;
-
-        self.vtl = self
-            .get_register(hvdef::HvAllArchRegisterName::VsmVpStatus.into())
-            .map_or(Vtl::Vtl0, |status| {
-                hvdef::HvRegisterVsmVpStatus::from(status.as_u64())
-                    .active_vtl()
-                    .try_into()
-                    .unwrap()
-            });
     }
 
     /// Call before jumping to kernel.
-    pub fn uninitialize(&mut self) {
+    pub fn uninitialize(&mut self, isolation_type: IsolationType) {
         if self.initialized {
-            crate::arch::hypercall::uninitialize();
-            self.initialized = false;
+            match isolation_type {
+                #[cfg(target_arch = "x86_64")]
+                IsolationType::Tdx => {
+                    self.uninitialize_tdx();
+                }
+                #[cfg(target_arch = "x86_64")]
+                IsolationType::Snp => (),
+                _ => {
+                    crate::arch::hypercall::uninitialize();
+                }
+            }
         }
+        self.initialized = false;
     }
 
     /// Returns the environment's VTL.
-    pub fn vtl(&self) -> Vtl {
+    pub fn vtl(&mut self) -> Vtl {
```

Comment:
why is this mut? this shouldn't be mut, nor should we query the vtl every time we need it. just do it during init like before for non-isolated, but on isolated platforms, we hardcode VTL2.